File: learning-javascript-data-structures-and-algorithms/stacks/calPoints.js
import { Stack } from "./stack"

const calPoints = (ops) => {
  const stack = new Stack();
  for (let i = 0; i < ops.length; i++) {
    const op = ops[i];
    if (op === "+") {
      const top = stack.pop();
      const newTop = top + stack.peek();
      stack.push(top || 0);
      stack.push(newTop || 0);
    } else if (op === "D") {
      const top = stack.peek();
      const d = top * 2 || 0;
      stack.push(d);
    } else if (op === "C") {
      stack.pop();
    } else {
      stack.push(parseInt(op) || 0);
    }
  }
  let sum = 0;
  while (!stack.isEmpty()) {
    sum += stack.pop();
  }
  return sum;
}

export {
  calPoints
}

File: learning-javascript-data-structures-and-algorithms/stacks/calPoints.rs
use std::collections::VecDeque;

fn cal_points(ops: Vec<&str>) -> i32 {
    let mut stack: VecDeque<i32> = VecDeque::new();
    for op in ops {
        match op {
            "+" => {
                let top = *stack.back().unwrap_or(&0);
                let new_top = top + *stack.get(stack.len() - 2).unwrap_or(&0);
                stack.push_back(new_top);
            },
            "D" => {
                let top = *stack.back().unwrap_or(&0);
                let d = top * 2;
                stack.push_back(d);
            },
            "C" => { stack.pop_back(); },
            _ => {
                let num = op.parse::<i32>().unwrap_or(0);
                stack.push_back(num);
            }
        }
    }
    let sum: i32 = stack.iter().sum();
    sum
}